### ISSUE TITLE
blocking of inline-scripts in browser (either through ublock or CSP) breaks ublock's script injections

### ISSUE THREAD

**Issue by berrythesoftwarecodeprogrammar** [gorhill/uBlock#1516]
**maybe not an acceptable issue, but PoC below of breaking ublock's script injection**

Firefox 45.0.1
uBlock Origin 1.6.4

i was thinking about ublock and some other addons i use which rely on injecting javascript to change webpages / prevent bad things from happening, and i was wondering if using ublock's "block inline scripts" feature could break them, and also if a `Content-Security-Policy` header sent by a website which told the browser not to load inline-scripts could also break them.

what i did was create an html page which loads a javascript file which does the following:
1. write the useragent retrieved by javascript to the document
2. write `document.eval` and `window.eval` to the document
3. imitate an Ad function which ublock currently blocks. these lines in `filters.txt`:
```
# https://adblockplus.org/forum/viewtopic.php?f=2&t=25216#p149739
lequipe.fr##script:inject(sas-defuser.js)
```
(i looked at `resources.txt` and made my own Ads/ads function which does `alert(1)`.

the reason i did these 3 things was to see if CSP/blocking inline scripts would break:
1. my useragent spoofing addons
2. my noscript surrogate which turns the `eval` function into `null` (`window.eval = null;document.eval=null;`)
3. ublock's script injection which breaks an ads/Ads function

it turned out that it didnt break the first two, but it DOES break ublock's script injection.
below are 2 links to the same html+js except one has a CSP header (which blocks inline-js) and one without any CSP header. the javascript should `alert(1)` in your browser on both links (meaning the "ad" script loaded). add this filter to custom filters: `berr.yt##script:inject(sas-defuser.js)`. that will load the resource you already have to break that function. then reload both of my pages. you will see it successfully prevent the `alert(1)` from happening now on the non-csp page. but on the csp page, the website tells the browser to not run any inline-scripts, and then that breaks ublock's injection and runs the "ad" script anyway and gives you the alert. the same can also be achieved by disabling inline scripts in ublock. it prevents the injectors which break ads from loading, so the ads continue to load.

test links: (make sure to use the filter `berr.yt##script:inject(sas-defuser.js)`):
no CSP: [https://berr.yt/h/test/nocsp/linked-ua.htm](https://berr.yt/h/test/nocsp/linked-ua.htm)
with CSP: [https://berr.yt/h/test/csp/linked-ua.htm](https://berr.yt/h/test/csp/linked-ua.htm)

i dont know if any websites use these tricks yet but they could. is it possible at all to make the script injection/replacement done by ublock happen the way it does with noscript's surrogates? i dont know everything about how these things work but i guess the noscript surrogates may not be "injections" and may somehow do things before the page is loaded. and maybe ublock has no choice but to inject afterwards? i dont know. but the CSP/blocking of inline scripts doesnt break noscript's surrogates, only ublock's injections.

and as for ublock's "block inline scripts" option, maybe it should be made clear to users that if javascript is enabled for a website but inline scripts are blocked, its possible ads/other things which ublock blocks, COULD get through as javascript will load them but ublock's inline-script-blocking will prevent ublock's injections from kicking in and killing the bad javascript.

maybe none of this is that important i dont know. or maybe ublock cant fix it. just thought i should let you know. i hope it was clear enough

**Comment by berrythesoftwarecodeprogrammar** [comment]
oh and the CSP header i added for the csp page was `default-src 'none'; script-src 'self';`

**Comment by berrythesoftwarecodeprogrammar** [comment]
tl;dr
1. users should maybe know that if they allow javascript but block inline-scripts in ublock, then they risk blocking some of ublock's injections
2. websites can use CSP headers to break ublock's injections

**Comment by berrythesoftwarecodeprogrammar** [comment]
or you can test on your own servers:
**x.html**
```
<html>
<body>
</body>
<script type="text/javascript" src="ua.js"></script>
</html>
```

**ua.js**
```
var noopfn = function() {
	alert(1);
};
var ads = {
	display: noopfn,
	refresh: noopfn
};
Object.defineProperty(window, 'Ads', {
	get: function() {
		return ads;
	},
	set: noopfn
});
window.Ads.display();
```

**ublock filter**:
```
your.server##script:inject(sas-defuser.js)
```

and try with and without this CSP header:
```
default-src 'none'; script-src 'self';
```

**Comment by RoxKilly** [comment]
@berrythesoftwarecodeprogrammar  I was able to replicate your interesting findings: The `berr.yt##script:inject(sas-defuser.js)` injection fails to block the dummy-ad when the `content-security-policy` header is set.  I would be surprised if this were a popular technique to circumvent adblocking because this header disables inline scripts whereas most sites that load ads need inline scripting.

What I found a lot more interesting was your second observation.  When inline-scripts are blocked in uBO, the script injection rule `berr.yt##script:inject(sas-defuser.js)` fails so the dummy-ad is able to run from the external script.  This has implications for advanced users that I had never considered.  When trying to defeat a site, allowing inline-script may actually be necessary to block malicious external scripts.

Thanks for sharing. Cool findings.
